{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "296fe46f-9449-404d-a618-d85f08341271",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "This notebook has been created as an addition to the main pipeline as is not necessarily part of the analysis process. Instead it takes the raw data provided and goes throught the process or shaping it to be ready to be imported into euler for the analytical pipeline. Chronologically, this notebook sits before ***01_Data_Exploration***. \n",
    "\n",
    "the files we recieved are:\n",
    "- *Illumina_MAGs.tar.gz*\n",
    "- *PacBio_MAGs.tar.gz*\n",
    "- *merged_metadata_filtered.tsv*\n",
    "\n",
    "The samples provided have been sequenced with each technique, meaning that intuitively all samples for a technique have a duplicate counterpart for the other. Feature files in *Illumina_MAGs* look to be already in [UUIDv4]('https://moshpit.qiime2.org/en/stable/chapters/howtos/import/') naming convention, but *PacBio_MAGs* are not.\n",
    "\n",
    "what is missing and the changes that need to be made are:\n",
    "- Rename features with the [UUIDv4]('https://moshpit.qiime2.org/en/stable/chapters/howtos/import/')\n",
    "- Create a primary key for all samples to retain technique information.\n",
    "- Update the metadata file to reflect the new primary key.\n",
    "- Create a metadata file feature-wide for preliminary analisis.\n",
    "- Fullfill qiime2 requirments for import\n",
    "    - Merge the 2 datasets in a sample_data common directory so that it is qiime-compatible.\n",
    "    - Create a MANIFEST file for **MAGS** in the sample_data directory.\n",
    "    - Clean sample_data of possible .tmp files that would interfere with qiime compatibility.\n",
    "- Compress sample_data to be exported to the Euler cluster for analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a192da93-c003-4662-ba9e-4d543e268b57",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48de6ba4-287c-40d3-ad81-a5db15b27132",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#set up environment\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "from qiime2 import Visualization\n",
    "\n",
    "# create directories for the notebook. DO NOT change\n",
    "!mkdir -p data\n",
    "!mkdir -p data/processed\n",
    "raw_data = 'data/raw'\n",
    "metadata_dir = 'data/processed/metadata'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f243fc8d-c80f-4a33-85c8-d469e5e7fb6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Just for this once it's better to erase the raw_data folder if you already ran the notebook and intend to run it again\n",
    "!rm -r $raw_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0275ed82-9059-4d89-bd40-8708e94010f5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash -s \"$raw_data\" \"$metadata_dir\"\n",
    "mkdir -p \"$1\"\n",
    "mkdir -p \"$2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba38378-e6a5-4ac1-9382-f36ebe5184e4",
   "metadata": {},
   "source": [
    "# Downloading and unzipping raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db04ecac-86e2-4bd7-8464-84cdec643e9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-12-10 16:03:52--  https://polybox.ethz.ch/index.php/s/56JaAiKdGwioBKN/download\n",
      "Resolving polybox.ethz.ch (polybox.ethz.ch)... 129.132.71.243\n",
      "Connecting to polybox.ethz.ch (polybox.ethz.ch)|129.132.71.243|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [application/zip]\n",
      "Saving to: ‘additional_data/Download.zip’\n",
      "\n",
      "additional_data/Dow     [       <=>          ] 492.28M   371MB/s    in 1.3s    \n",
      "\n",
      "2025-12-10 16:03:54 (371 MB/s) - ‘additional_data/Download.zip’ saved [516188455]\n",
      "\n",
      "Archive:  additional_data/Download.zip\n",
      "   creating: data/raw/applied_bioinformatics/\n",
      " extracting: data/raw/applied_bioinformatics/.DS_Store  \n",
      " extracting: data/raw/applied_bioinformatics/Illumina_MAGs.tar.gz  \n",
      " extracting: data/raw/applied_bioinformatics/PacBio_MAGs.tar.gz  \n",
      " extracting: data/raw/applied_bioinformatics/merged_metadata_filtered.tsv  \n"
     ]
    }
   ],
   "source": [
    "# Download files from the polybox\n",
    "!wget 'https://polybox.ethz.ch/index.php/s/56JaAiKdGwioBKN/download'  -O $raw_data/Download.zip\n",
    "# Unzip\n",
    "!unzip -o $raw_data/Download.zip -d $raw_data\n",
    "!rm $raw_data/Download.zip\n",
    "# extract contents from applied_bioinformatics folder\n",
    "!mv $raw_data/applied_bioinformatics/* $raw_data\n",
    "!rm -r $raw_data/applied_bioinformatics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596599c4-9a6a-42f7-bac2-84e0b44e4474",
   "metadata": {
    "tags": []
   },
   "source": [
    "## .tar.gz\n",
    "the MAGs are stored in .tar.gz files for Illumina- and PacBio- generated sequences. Extracting the files from the is the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a7137901-123a-4a2f-b05d-81dc6145c263",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#untar MAGs files and store everything in raw_data\n",
    "!tar -xzf $raw_data/Illumina_MAGs.tar.gz -C $raw_data\n",
    "!tar -xzf $raw_data/PacBio_MAGs.tar.gz -C $raw_data\n",
    "#remove the .tar.gz files\n",
    "!rm -r $raw_data/Illumina_MAGs.tar.gz\n",
    "!rm -r $raw_data/PacBio_MAGs.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78f91c0-39db-4457-8183-d5ea9a92b677",
   "metadata": {},
   "source": [
    "<a id='UUID'></a>\n",
    "## UUIDv4 renaming\n",
    "A simple nested for loop that goes throught every file in the directories of interest and applies the uuid as the **new_path**. UUIDv4 is a 32 digit hexadecimal string that is semi-randomly generated. It's used to ensure that feature names are unique and 32 digits (128bit) long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fd6cb744-022a-46e5-962f-2435557643d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from uuid import uuid4\n",
    "\n",
    "#Rename all fasta files with UUIDs\n",
    "for technique in os.listdir(raw_data):\n",
    "    tech_path = os.path.join(raw_data, technique)\n",
    "    if not os.path.isdir(tech_path):\n",
    "        continue\n",
    "    for sample_id in os.listdir(tech_path):\n",
    "        sample_path = os.path.join(tech_path, sample_id)\n",
    "        if not os.path.isdir(sample_path):\n",
    "            continue\n",
    "        for file in os.listdir(sample_path):\n",
    "            if file.endswith((\".fa\", \".fasta\")):\n",
    "                old_path = os.path.join(sample_path, file)\n",
    "                new_path = os.path.join(sample_path, f\"{uuid4()}.fa\")\n",
    "                os.rename(old_path, new_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c6e0c9-bae2-4377-9aff-bb6860f04875",
   "metadata": {},
   "source": [
    "<a id='UUID'></a>\n",
    "\n",
    "## Primary Key Generation\n",
    "By primary key, a unequivocal ID is intended. Right now if we were to merge the features and look at them with the metadata information, the MAGs coming from the same sample and sequenced with Illumina would not be distinguishable from the Pacbio ones and vice versa. As primary key, the a technique dependent prefix ('IL' for Illumina and 'PB' for PacBio) is enough to discern.\n",
    "to do so we will use a for loop again, adding prefix + '_' to the name of each sample directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86fe0bb6-a40d-486a-8208-1d01a4bdd2e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash -s \"$raw_data\"\n",
    "raw_data=\"$1\"\n",
    "# Rename all samples in Illumina_MAGs \n",
    "for d in $raw_data/Illumina_MAGs/*; do\n",
    "    mv \"$d\" \"$(dirname \"$d\")/IL_$(basename \"$d\")\"\n",
    "done\n",
    "for d in $raw_data/PacBio_MAGs/*; do\n",
    "    mv \"$d\" \"$(dirname \"$d\")/PB_$(basename \"$d\")\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4481f89b-b332-4a80-9eaa-4dab0db1115a",
   "metadata": {},
   "source": [
    "It is important that the new primary key is included in the metadata file, the following code will retreive the new sample name. Additionally it will retrieve the filename of the fasta files. This is not required for the primary key, but will make it easier to perform Data Exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a796e613-1e09-40fb-9ec8-8c676d97f801",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import metadata\n",
    "metadata_df = pd.read_csv(f\"{raw_data}/merged_metadata_filtered.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "#\n",
    "prefixes = (\"PB_\", \"IL_\")\n",
    "records = []\n",
    "\n",
    "for technique in os.listdir(raw_data):\n",
    "    tech_path = os.path.join(raw_data, technique)\n",
    "    if not os.path.isdir(tech_path):\n",
    "        continue\n",
    "    for sample_id in os.listdir(tech_path):\n",
    "        sample_path = os.path.join(tech_path, sample_id)\n",
    "        if not os.path.isdir(sample_path):\n",
    "            continue\n",
    "\n",
    "        # derive ID for lookup\n",
    "        lookup_id = sample_id\n",
    "        for p in prefixes:\n",
    "            if lookup_id.startswith(p):\n",
    "                lookup_id = lookup_id[len(p):]\n",
    "\n",
    "        for f in os.listdir(sample_path):\n",
    "            if f.endswith((\".fa\", \".fasta\")):\n",
    "                if lookup_id in metadata_df.index:\n",
    "                    mag_id = os.path.splitext(f)[0]\n",
    "                    records.append((sample_id, mag_id))\n",
    "\n",
    "primary_df = pd.DataFrame(records, columns=[\"sample-id\", \"mag-id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "991014ee-e276-43af-819d-47d6edf70ff4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['PB_B039_Aa_Gp_La', 'PB_MS001-3', 'PB_9b8b5', 'PB_M008',\n",
       "       'PB_B038_Az_Gp_La', 'PB_e7c76', 'PB_P009', 'PB_M010',\n",
       "       'PB_HM010-03', 'PB_B056_Sc_Na_Af', 'PB_HM010-01', 'PB_M004',\n",
       "       'PB_36fe4', 'PB_a36ba', 'PB_M009', 'PB_P003', 'PB_P001', 'PB_M012',\n",
       "       'PB_A001', 'PB_MS013-1', 'PB_B051_Aj_Po_Laf', 'PB_MS003-2',\n",
       "       'PB_M006', 'PB_MS005-1', 'PB_B037_La_Ac_La', 'PB_MS009-1',\n",
       "       'PB_3ee22', 'IL_B038_Az_Gp_La', 'IL_A001', 'IL_P003', 'IL_MS003-3',\n",
       "       'IL_M009', 'IL_HM010-03', 'IL_M010', 'IL_B044_Hb_Ac_Ab',\n",
       "       'IL_MS009-2', 'IL_MS009-1', 'IL_MS001-3', 'IL_M002', 'IL_M012',\n",
       "       'IL_MS011-1', 'IL_B056_Sc_Na_Af', 'IL_M008', 'IL_B037_La_Ac_La',\n",
       "       'IL_MS013-1', 'IL_MS005-1', 'IL_MS003-2', 'IL_HM010-01', 'IL_P009',\n",
       "       'IL_B051_Aj_Po_Laf', 'IL_A002', 'IL_B039_Aa_Gp_La'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check sample-id\n",
    "primary_df['sample-id'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26e9434d-7822-4201-b377-fe15d7e95d1a",
   "metadata": {},
   "source": [
    "Now that everybody is accounted for, we can derive the old sample name to merge the new information to the metadata file provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3c6d5b29-1447-43ee-a444-4accfed7ef7f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>mag-id</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>6de933c6-afeb-4e58-b764-f8ba5fc0d366</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>8d28fd5f-5f2c-4508-9764-1c20ebffa8df</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>c0a09ab4-2630-4493-a84e-67f8b764d4d9</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample-id                                mag-id         sample\n",
       "0  PB_B039_Aa_Gp_La  6de933c6-afeb-4e58-b764-f8ba5fc0d366  B039_Aa_Gp_La\n",
       "1  PB_B039_Aa_Gp_La  8d28fd5f-5f2c-4508-9764-1c20ebffa8df  B039_Aa_Gp_La\n",
       "2  PB_B039_Aa_Gp_La  c0a09ab4-2630-4493-a84e-67f8b764d4d9  B039_Aa_Gp_La"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create 'sample' without the IL/PB prefix\n",
    "primary_df['sample'] = primary_df['sample-id'].str.split(pat='_', n=1).str[1]\n",
    "primary_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d115639-6bff-410a-af90-d284aac93e5b",
   "metadata": {},
   "source": [
    "## feature-based metadata\n",
    "for distributions and preliminary analisis a count of of fasta per could be useful. We merge primary_df with metadata on the left to achieve this result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3e81474e-0513-423c-b1cf-53d7e08f16ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>mag-id</th>\n",
       "      <th>sample</th>\n",
       "      <th>samp_country</th>\n",
       "      <th>category</th>\n",
       "      <th>fermented_food_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>6de933c6-afeb-4e58-b764-f8ba5fc0d366</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>8d28fd5f-5f2c-4508-9764-1c20ebffa8df</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>c0a09ab4-2630-4493-a84e-67f8b764d4d9</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample-id                                mag-id         sample  \\\n",
       "0  PB_B039_Aa_Gp_La  6de933c6-afeb-4e58-b764-f8ba5fc0d366  B039_Aa_Gp_La   \n",
       "1  PB_B039_Aa_Gp_La  8d28fd5f-5f2c-4508-9764-1c20ebffa8df  B039_Aa_Gp_La   \n",
       "2  PB_B039_Aa_Gp_La  c0a09ab4-2630-4493-a84e-67f8b764d4d9  B039_Aa_Gp_La   \n",
       "\n",
       "  samp_country        category fermented_food_type  \n",
       "0        Benin  fermented fish            Lanhouin  \n",
       "1        Benin  fermented fish            Lanhouin  \n",
       "2        Benin  fermented fish            Lanhouin  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = primary_df.merge(metadata_df, left_on ='sample', right_index = True, how = 'left')\n",
    "merged_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b5f03d-04fc-4730-a9fa-3b230e6edb6b",
   "metadata": {},
   "source": [
    "A column with the relative path for jupyter will be added to make Data exploration easier. the file path can be made with *raw_data/sample_data/* and *sample-id/mag-id* from the merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bc623f29-12fd-49ea-b0f4-6239550b84b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>mag-id</th>\n",
       "      <th>sample</th>\n",
       "      <th>samp_country</th>\n",
       "      <th>category</th>\n",
       "      <th>fermented_food_type</th>\n",
       "      <th>file_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>6de933c6-afeb-4e58-b764-f8ba5fc0d366</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "      <td>data/raw/sample_data/PB_B039_Aa_Gp_La/6de933c6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>8d28fd5f-5f2c-4508-9764-1c20ebffa8df</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "      <td>data/raw/sample_data/PB_B039_Aa_Gp_La/8d28fd5f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>c0a09ab4-2630-4493-a84e-67f8b764d4d9</td>\n",
       "      <td>B039_Aa_Gp_La</td>\n",
       "      <td>Benin</td>\n",
       "      <td>fermented fish</td>\n",
       "      <td>Lanhouin</td>\n",
       "      <td>data/raw/sample_data/PB_B039_Aa_Gp_La/c0a09ab4...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample-id                                mag-id         sample  \\\n",
       "0  PB_B039_Aa_Gp_La  6de933c6-afeb-4e58-b764-f8ba5fc0d366  B039_Aa_Gp_La   \n",
       "1  PB_B039_Aa_Gp_La  8d28fd5f-5f2c-4508-9764-1c20ebffa8df  B039_Aa_Gp_La   \n",
       "2  PB_B039_Aa_Gp_La  c0a09ab4-2630-4493-a84e-67f8b764d4d9  B039_Aa_Gp_La   \n",
       "\n",
       "  samp_country        category fermented_food_type  \\\n",
       "0        Benin  fermented fish            Lanhouin   \n",
       "1        Benin  fermented fish            Lanhouin   \n",
       "2        Benin  fermented fish            Lanhouin   \n",
       "\n",
       "                                           file_path  \n",
       "0  data/raw/sample_data/PB_B039_Aa_Gp_La/6de933c6...  \n",
       "1  data/raw/sample_data/PB_B039_Aa_Gp_La/8d28fd5f...  \n",
       "2  data/raw/sample_data/PB_B039_Aa_Gp_La/c0a09ab4...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[\"file_path\"] = merged_df.apply(\n",
    "    lambda r: f\"{raw_data}/sample_data/{r['sample-id']}/{r['mag-id']}\",\n",
    "    axis=1\n",
    ")\n",
    "merged_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b92a27e-6c56-4391-8b68-ff082251d083",
   "metadata": {},
   "source": [
    "The new *merged_df* can be saved as *Metadata_Extended.tsv* and used for exploratory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "589e9ae5-fa6d-49e9-b7bb-da0d95dc30f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "merged_df.to_csv(f\"{metadata_dir}/Metadata_Extended.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a09657ac-fc90-4c0c-aa82-577c588bc962",
   "metadata": {},
   "source": [
    "## Getting things ready for qiime2 Import\n",
    "### fixing directory tree\n",
    "To import sequences correctly into qiime we need to modify the directory tree to match from the current:\n",
    "\n",
    "data/  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;- technique 1  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;- technique2  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- sample 1  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- sample 2  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- seq1.fa  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- seq2.fa  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...  \n",
    "To the required:  \n",
    "data/  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;sample1  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;sample2  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-seq1  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-seq2  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...  \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;MANIFEST  \n",
    "\n",
    "qiime would import the original tree without issue, but the subsequent BUSCO Evaluation job would fail. The error was due to some issue with the pathing.\n",
    "\n",
    "Let's move all the sample folders found to a new *raw_data/sample_data* folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "35e758dd-938c-4f01-ba87-e2bae6b7464b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash -s \"$raw_data\"\n",
    "raw_data=\"$1\"\n",
    "\n",
    "# Destination folder\n",
    "dest_dir=\"$raw_data/sample_data\"\n",
    "mkdir -p \"$dest_dir\"\n",
    "\n",
    "mv \"$raw_data/Illumina_MAGs\"/* \"$dest_dir\"/\n",
    "mv \"$raw_data/PacBio_MAGs\"/* \"$dest_dir\"/\n",
    "\n",
    "# Delete the directory husk left behind\n",
    "rm -r $raw_data/Illumina_MAGs\n",
    "rm -r $raw_data/PacBio_MAGs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13afdc1f-de6a-4d10-ada7-3a0300530c13",
   "metadata": {},
   "source": [
    "### MANIFEST"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c7900a-6206-413c-bbad-df1b19f5acff",
   "metadata": {},
   "source": [
    "It is a file that associates each files data to their location. By following the [Moshpit Tutorial]('https://moshpit.qiime2.org/en/stable/chapters/howtos/import/') for importing non-dereplicated mags, no information about the MANIFEST format is provided. This is not a problem since the format was taken from the week 3 tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "098155cb-5d5d-4392-be6a-f7b7caa0972e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-12-10 16:04:58--  https://polybox.ethz.ch/index.php/s/LBs9m5ZAAxGTLR6/download\n",
      "Resolving polybox.ethz.ch (polybox.ethz.ch)... 129.132.71.243\n",
      "Connecting to polybox.ethz.ch (polybox.ethz.ch)|129.132.71.243|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3012 (2.9K) [application/octet-stream]\n",
      "Saving to: ‘data/raw/w3_MANIFEST’\n",
      "\n",
      "data/raw/w3_MANIFES 100%[===================>]   2.94K  --.-KB/s    in 0s      \n",
      "\n",
      "2025-12-10 16:04:58 (549 MB/s) - ‘data/raw/w3_MANIFEST’ saved [3012/3012]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>absolute-filepath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>550.L1S109.s.1</td>\n",
       "      <td>$PWD/w3_data/seqs/ERR1866491.fastq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample-id                   absolute-filepath\n",
       "0  550.L1S109.s.1  $PWD/w3_data/seqs/ERR1866491.fastq"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import MANIFEST from the tutorial\n",
    "!wget 'https://polybox.ethz.ch/index.php/s/LBs9m5ZAAxGTLR6/download' -O $raw_data/w3_MANIFEST\n",
    "\n",
    "w3 = pd.read_csv(f'{raw_data}/w3_MANIFEST', sep = '\\t')\n",
    "w3.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fd4658c-120d-42e4-b953-0424f7926bc8",
   "metadata": {},
   "source": [
    "As you can be seen the MANIFEST needs to be a **2 column**, **tab separated** file to be compatible with qiime. From the Moshpit tutorial we know that the sequences files names have to be UUIDv4 generated. It should follow naturally that for the mags a **3 column, comma separated value** is required."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc684b9b-184a-48fc-897f-80975aac1131",
   "metadata": {
    "tags": []
   },
   "source": [
    "It is a matter of creating a df with **'sample-id'** and **'mags-id'** from our newly created *merged_df* and the **path** of each file in *sample_data*, making sure that the path matches the future location in the cluster, as they need to be imported in a cache for ease of access of the computational nodes. This has actually been one of the first roadblocks for the project as understanding how to get everything in ready-to-go state on Jupyter (where the UI is more visually intuitive for novice bioinformaticians) and move it to Euler was a real challenge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "29519428-4e2f-4577-8e7a-ff3efe541eae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample-id</th>\n",
       "      <th>mag-id</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>6de933c6-afeb-4e58-b764-f8ba5fc0d366</td>\n",
       "      <td>/cluster/scratch/emotta/sample_data/PB_B039_Aa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>8d28fd5f-5f2c-4508-9764-1c20ebffa8df</td>\n",
       "      <td>/cluster/scratch/emotta/sample_data/PB_B039_Aa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB_B039_Aa_Gp_La</td>\n",
       "      <td>c0a09ab4-2630-4493-a84e-67f8b764d4d9</td>\n",
       "      <td>/cluster/scratch/emotta/sample_data/PB_B039_Aa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample-id                                mag-id  \\\n",
       "0  PB_B039_Aa_Gp_La  6de933c6-afeb-4e58-b764-f8ba5fc0d366   \n",
       "1  PB_B039_Aa_Gp_La  8d28fd5f-5f2c-4508-9764-1c20ebffa8df   \n",
       "2  PB_B039_Aa_Gp_La  c0a09ab4-2630-4493-a84e-67f8b764d4d9   \n",
       "\n",
       "                                            filename  \n",
       "0  /cluster/scratch/emotta/sample_data/PB_B039_Aa...  \n",
       "1  /cluster/scratch/emotta/sample_data/PB_B039_Aa...  \n",
       "2  /cluster/scratch/emotta/sample_data/PB_B039_Aa...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure to change your \n",
    "Euler_dir = '/cluster/scratch/emotta/sample_data'\n",
    "\n",
    "manifest_df = merged_df[[\"sample-id\", \"mag-id\"]].copy()\n",
    "manifest_df[\"filename\"] = manifest_df.apply(\n",
    "    lambda r: f\"{Euler_dir}/{r['sample-id']}/{r['mag-id']}.fa\",\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "manifest_df.head(3)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2848fb-7e34-469b-80a7-d083e5a01848",
   "metadata": {},
   "source": [
    " This has easily been one of the most challenging steps in our project, since no proper documentation exist about this and even asking the the wednesday's session supervisors, the only guide was the error messages from the ***Mosh tools cache-import*** command when trying to import the sequences in qiime.\n",
    " \n",
    " <span style=\"color:red\">*MANIFEST is not a(n) MultiMAGManifestFormat file  \n",
    "Found header on line 1 with the following labels: ['sample-id\\tabsolute-filepath'], expected: ['sample-id', 'mag-id', 'filename']*</span>\n",
    "\n",
    "And later \n",
    "\n",
    "<span style=\"color:red\"> Found header on line 1 with the following labels: ['sample-id\\tmag-id\\tfilename'], expected: ['sample-id', 'mag-id', 'filename']</span>\n",
    "\n",
    "In retrospective it was not a complicated issue, but summed with the steep learning curve of both qiime and Euler simultaneously it added up to a great setback of the first part of the semester.\n",
    "\n",
    "We are now ready to save the MANIFEST in the *sample_data* directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9b1c8e9e-c79a-43c2-9599-c4104378be2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "manifest_df.to_csv(f'{raw_data}/sample_data/MANIFEST', sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3f114ae-b151-4d7d-82e6-49d0bf58986f",
   "metadata": {},
   "source": [
    "## Clean sample_data\n",
    "another common issue we encountered was the error warning was due to the presence of temporary and hidden files (*.files*) created by jupyter and moved from the raw data directories.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "16bcd57b-bcfa-469d-81f9-17482299e431",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!find \"$raw_data/sample_data\" -maxdepth 1 -type f -name \".*\" -delete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2e183ed-73c8-46a2-b953-061dc23c3214",
   "metadata": {},
   "source": [
    "## Zip and upload to Euler\n",
    "*sample_data* has been "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e38fec51-6640-4683-aa72-84780d3759b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tar: sample_data: file changed as we read it\n"
     ]
    }
   ],
   "source": [
    "!tar -czf sample_data.tar.gz -C \"$raw_data\" sample_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa8dccfa-dafe-437a-945c-c505b507e569",
   "metadata": {},
   "source": [
    "**all set!** *sample_data* is now ready to be downloaded and manually uploaded to the ${HOME}/applied_bioinformatics/03_Aritifacts_Zipped directory on Euler for safe keeping"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "QIIME 2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
